File: 3058-maximum-number-of-k-divisible-components/3058-maximum-number-of-k-divisible-components.cpp
class Solution {
public:
    int dfs(int curr, int prev, vector<vector<int>>& adj, vector<int>& values, int k, int& maxComp){
        
        int curr_sum = values[curr];
        
        for(int nei : adj[curr]){
            if(nei != prev){
                curr_sum += dfs(nei,curr,adj,values,k,maxComp) % k;
            }
        }
        
        if(curr_sum % k == 0){
            maxComp += 1;
            return 0;
        }

        return curr_sum;
    }

    int maxKDivisibleComponents(int n, vector<vector<int>>& edges, vector<int>& values, int k) {
        vector<vector<int>> adj(n);
        vector<bool> vis(n,false);
        for(auto edge : edges){
            adj[edge[0]].push_back(edge[1]);
            adj[edge[1]].push_back(edge[0]);
        }

        int maxComp = 0;

        dfs(0,-1,adj,values,k,maxComp);

        return maxComp;
    }
};

File: 3058-maximum-number-of-k-divisible-components/3058-maximum-number-of-k-divisible-components.py
class Solution:
    def maxKDivisibleComponents(self, n: int, edges: List[List[int]], values: List[int], k: int) -> int:
        adj = [[] for _ in range(n+1)] 

        for edge in edges:
            adj[edge[0]].append(edge[1])
            adj[edge[1]].append(edge[0])
        
        ans = [0]

        def func(curr, parent):
            
            node_sum = values[curr]

            for child in adj[curr]:
                if child == parent:
                    continue
                node_sum += func(child,curr)
            
            if node_sum % k == 0:
                ans[0] += 1
                return 0
            
            return node_sum

        func(0,-1)

        return ans[0]